### ISSUE TITLE
Slow incremental linking times for fastOptJS

### ISSUE THREAD

**Issue by olafurpg** [scala-js/scala-js#2827]
Steps to reproduce

- clone https://github.com/olafurpg/scalameta/tree/scala.js-slow-fastopjs
- make sure you are on commit https://github.com/olafurpg/scalameta/commit/f1d0149c4cd061ca13854ed5475e7bbb3d8a255a
- `sbt scalametaJS/test:fastOptJS` (takes a long time, as expected)
- Add blank line **inside a `test` block** in PublicSuite, for example line 18
- `sbt scalametaJS/test:fastOptJS` (takes ~16 seconds to optimize 9 methods, unexpectedly long)

```
[scala.js-slow-fastopjs] scalameta: last scalametaJS/test:fastOptJS
[info] Fast optimizing /Users/ollie/dev/scalameta/scalameta/scalameta/js/target/scala-2.11/scalameta-test-fastopt.js
[debug] Linker: Compute reachability: 2365588 us
[debug] Linker: Assemble LinkedClasses: 481434 us
[debug] Basic Linking: 2868037 us
[debug] Inc. optimizer: Batch mode: false
[debug] Inc. optimizer: Incremental part: 303286 us
[debug] Inc. optimizer: Optimizing 9 methods.
[debug] Inc. optimizer: Optimizer part: 16756267 us
[debug] Inc. optimizer: 17319874 us
[debug] Refiner: Compute reachability: 1456534 us
[debug] Refiner: Assemble LinkedClasses: 65163 us
[debug] Refiner: 1521887 us
[debug] Emitter: Class tree cache stats: reused: 6059 -- invalidated: 3
[debug] Emitter: Method tree cache stats: resued: 15877 -- invalidated: 7
[debug] Emitter (write output): 1220303 us
[debug] Global IR cache stats: reused: 3575 -- invalidated: 3 -- trees read: 3
```

Hardware: 
- MacBook Pro (Retina, 15-inch, Mid 2014)
- Processor: 2,5 GHz Intel Core i7
- Memory: 16 GB 1600 MHz DDR3

**Comment by olafurpg** [comment]
Here's the results from a cleaner fastOptJS (26k methods). Note that the "Inc. optimizer: Optimizer part" takes close to 5 minutes.

```
[scala.js] scalameta: last scalametaJS/test:fastOptJS
[info] Fast optimizing /Users/ollie/dev/scalameta/scalameta/scalameta/js/target/scala-2.11/scalameta-test-fastopt.js
[debug] Linker: Compute reachability: 1740008 us
[debug] Linker: Assemble LinkedClasses: 2826459 us
[debug] Basic Linking: 4592429 us
[debug] Inc. optimizer: Batch mode: true
[debug] Inc. optimizer: Incremental part: 546128 us
[debug] Inc. optimizer: Optimizing 26786 methods.
[debug] Inc. optimizer: Optimizer part: 286879693 us
[debug] Inc. optimizer: 288565146 us
[debug] Refiner: Compute reachability: 1993541 us
[debug] Refiner: Assemble LinkedClasses: 116146 us
[debug] Refiner: 2131236 us
[debug] Emitter: Class tree cache stats: reused: 3047 -- invalidated: 3015
[debug] Emitter: Method tree cache stats: resued: 0 -- invalidated: 15887
[debug] Emitter (write output): 7537673 us
[debug] Global IR cache stats: reused: 0 -- invalidated: 3578 -- trees read: 3526
```

**Comment by ngbinh** [comment]
this is also a huge issue for us when working on a large scala.js code base.

**Comment by olafurpg** [comment]
Are you using scalatest @ngbinh ? It could be that the slow linking times are caused by over-aggressive inlining of closures, see https://github.com/scalatest/scalatest/search?utf8=%E2%9C%93&q=%22testFun%3A+%3D%3E+Any%22&type=

**Comment by ngbinh** [comment]
ah, yes. We are using scalatest. This explains why our `test/fastOptJS` is usually very long. But even our source `fastOptJS` also usually takes more than 2 minutes with a simple change (our final output js file is around 60MB though)

**Comment by olafurpg** [comment]
I made a quick informal experiment to see if adding `@noinline` to scalatest helps.

Here is the diff I made to scalatest https://github.com/olafurpg/scalatest/commit/7822f5e110e0add9afd8212b336f9fea8dae8129

I published this change locally as scalatest 3.0.1-noinline and used it in https://github.com/scalameta/scalameta to run the task `scalametaJS/test:fastOptJS`

```
3.0.1 (without @noinline):
Clean [debug] Inc. optimizer: Optimizer part: 18346380 us
Edit  [debug] Inc. optimizer: Optimizer part: 4455692 us
File size: 16.7 MB

3.0.1-noinline (with @noinline):
Clean [debug] Inc. optimizer: Optimizer part: 17327262 us
Edit  [debug] Inc. optimizer: Optimizer part: 593117 us
File size: 16.3 MB
```

This benchmark methodology is not very rigorous, but the 29x difference in the "edit" part is interesting. I think these results at least suggest that over-inlining may be partly to blame for slow linking times.

Question is whether scalatest should add `@noinline` to common by-name parameters or if the Scala.js optimizer should be less aggressive.

Note. We are using https://scalacenter.github.io/scalajs-bundler/

**Comment by sjrd** [comment]
Since the Scala.js optimizer has no notion of "budget" for optimizations, the shortest path to get this fixed is by far to add the relevant `@noinline`s to ScalaTest.

**Comment by ngbinh** [comment]
@olafurpg interesting! The time difference definitely nice but I am surprised that the output file size difference is that small.

**Comment by ngbinh** [comment]
@olafurpg btw, is it possible to use the rewrites in olafurpg/scalatest@7822f5e on another `scala.js` code base? I would love to try it on our own :)

**Comment by olafurpg** [comment]
@ngbinh Sure thing!

Here is the rewrite and latest pre-release binary of `scalafix`: https://github.com/dotty-staging/scalatest/commit/42cd855138db82a7fe5d5efd1a5a77b9591ca02d

You can run it with: `./scalafix --rewrites file:rewrites/NoInline.scala -i scalatest/src/main/scala`

 Note, it may take a while to download everything on the first run. You can adapt NoInline.scala to your codebase and run scalafix again. There's little to no docs at the moment on the scalafix rewrite API, but you can take a look at some examples here https://github.com/scalacenter/scalafix/tree/master/scalafix-core/src/main/scala/scalafix/rewrite

**Comment by sjrd** [comment]
FWIW, #2964 seems to fix this problem to a large extent. On my machine, the cold `scalametaJS/test:fastOptJS` went from ~15 s down to 4.5 s. The incremental re-optimization after adding a blank line in `PublicSuite` line 18 was 2.3 s in "Inc. Optimizer", of which only ~0.1 s in "Optimizer part".

I still believe that the `@noinline`s can further improve the situation, but it is now clear that the Scala.js optimizer was to blame for its slowness.

**Comment by sjrd** [comment]
Optimistically closing as fixed by #2964.

**Comment by olafurpg** [comment]
I can confirm that 0.6.17 cuts linking times in the Scalameta test suite from ~17s to ~3s. I ran the command below

```
; scalametaJS/clean ; show scalametaJS/test:fastOptJS ; last scalametaJS/test:fastOptJS
```

six times with each of 0.6.16 and 0.6.17 and recorded the "Inc. optimizer" part. The [results](https://docs.google.com/spreadsheets/d/19CDsH-Tuc56-Seno-sxYvX4Z4GAPJ1qPjBTIdvPdd70/edit#gid=0) are in micro seconds

```
	0.6.16		0.6.17
stddev	4.08E+06	5.08E+05
avg	1.78E+07	2.95E+06
			6.02x faster
```

Thank you for the fix @sjrd ❤️
